Bluetooth: GATT: Skip certain attributes when discovering descriptors

Vudentz opened the pull request:
When discovering descriptor Find Information procedure is used which
does not allow any filtering by the server so it will return all
attributes in the given range including services and characteristics.

Fixes #14265

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Skip certain attributes when discovering descriptors

When discovering descriptor Find Information procedure is used which
does not allow any filtering by the server so it will return all
attributes in the given range including services and characteristics.

Fixes #14265

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

carlescufi commented on subsys/bluetooth/host/gatt.c:
```diff
@@ -2252,6 +2252,14 @@ static void gatt_find_info_rsp(struct bt_conn *conn, u8_t err,
 		}
 
+		/* Skip attributes that are not considered descriptors */
+		if (!bt_uuid_cmp(&u.uuid, BT_UUID_GATT_PRIMARY) ||
```
I assume there's no direct use of `Find Information Request/Response` by the application.
But what about characteristic values? those will still come through won't they? does this make sense?

Vudentz replied:
Well we can't possible filter all of the characteristics, that table would be too big besides that would have to include vendor one as well so I don't think it is doable, we could perhaps add some state tracking for the declaration and then skip the next handle.

carlescufi replied:
Right but is it worth it? Shouldn't the user know to pass the correct range when discovering descriptors?

carlescufi commented:
I am not convinced this is a bug when using Find Info Req/Rsp, because there is simply no way for the client to filter out non-descriptor attributes unless it knows the handle ranges

Vudentz commented:
> I am not convinced this is a bug when using Find Info Req/Rsp, because there is simply no way for the client to filter out non-descriptor attributes unless it knows the handle ranges

I went back and forth with the thought if that should be handle in the stack or not, in the end I think it is simpler to filter this on the stack rather then on the application thus why I submitted this anyway.

carlescufi commented:
> > I am not convinced this is a bug when using Find Info Req/Rsp, because there is simply no way for the client to filter out non-descriptor attributes unless it knows the handle ranges
> 
> I went back and forth with the thought if that should be handle in the stack or not, in the end I think it is simpler to filter this on the stack rather then on the application thus why I submitted this anyway.

Right, but there won't be any real way for the app to know whether the returned attributes are descriptors or not, correct? I think this merits at least a change in the doc reflecting this.

Vudentz commented:
> Right, but there won't be any real way for the app to know whether the returned attributes are descriptors or not, correct? I think this merits at least a change in the doc reflecting this.

Sure, I will update the documentation with the new behavior.

Vudentz committed:
Bluetooth: GATT: Skip Characteristic Value when discovering Descriptors

Those are not considered Descriptors and shall be discovered with
use of BT_GATT_DISCOVER_CHARACTERISTIC.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Document discover types

This documents existing types that can be used with bt_gatt_discover.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz committed:
Bluetooth: GATT: Add BT_GATT_DISCOVER_ATTRIBUTE type

This adds BT_GATT_DISCOVER_ATTRIBUTE which can be used to discover any
type of attribute in a given range.

Signed-off-by: Luiz Augusto von Dentz <luiz.von.dentz@intel.com>

Vudentz commented:
@carlescufi @paulgavrikov Ive added the documentation for the discover types and also include a new one that supports discovering any type of attribute.

carlescufi reviewed (approved):
Looks good now, thanks!

paulgavrikov commented:
Looks good
